### ISSUE TITLE
"constructor-style" type-case

### ISSUE THREAD

**Issue by stamourv** [racket/plai#6]
`type-case`'s clause syntax is as follows:
`[constructor-name (field ...) body]`

@rfindler observes (and I agree) that this indents poorly, and it's also inconsistent with, e.g., `match`.

This syntax, on the other hand, does not have these issues:
`[(constructor field ...) body]`

Of course, we can't just change `type-case` to accept the latter and not the former. Backwards compatibility, textbook, etc.

I see two other ways to support this syntax, both of which have pros and cons. It's not clear to me which (if any) would be a good idea.

1. Introduce a new form, say `type-case*` that is just like `type-case`, but with the new syntax. That wouldn't break existing code, but having two almost-but-not-quite identical forms always makes me sad. (I'm still sad when I think about `struct` / `define-struct`.)

2. Extend `type-case` to support both syntaxes. That sounds like it could cause some headaches re error messages, though.

Both approaches would also diverge from what's in the textbook, which may or may not be an issue.

Whatever we do, `typed-plai` should probably be changed in the same way.

**Comment by jeapostrophe** [comment]
I don't think that SK intends to update the book anytime, but perhaps he would let us.

If there enough things like this, then "#lang plai2" is good.

If not, then I think `type-case*` or `type-match` would be good.

**Comment by stamourv** [comment]
I don't have anything else of that size, but I haven't been using plai as long as many of you have.

If no one else has anything large enough to justify a plai2, then I'll add a `type-match` (I like your name better). And we can add it to the list for an eventual plai2 (along with gc2).

**Comment by jeapostrophe** [comment]
I don't have anything else (in fact, I have got to the point that I often tell the students to use `(require plai/datatype)` with `#lang racket` so they can trust random documentation more.
